lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,164 @@
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from xmodule_django.models import BlockTypeKeyField, CourseKeyField
+
+from hashlib import md5
+import json
+
+
+class VisibleBlocksModel(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=32, primary_key=True)
+
+    def __init__(self, *args, **kwargs):
+        """
+        Creates a new VisibleBlocksModel. Keyword argument 'blocks' should be an array of BlockRecords. KeyError will
+        be raised if this argument is not present.
+        """
+        blocks = kwargs.pop('blocks')
+        super(VisibleBlocksModel, self).__init__(*args, **kwargs)
+        self._blocks_json = json.dumps(
+            [
+                block.to_json()
+                for block in blocks
+            ]
+        )
+        self.hashed = md5(self._blocks_json).hexdigest()
```

Comment:
`hash()` wasn't trivial to track down, which is why I decided on md5. From StackOverflow, this appears to be the way it's implemented for strings in 2.x: http://svn.python.org/view/python/trunk/Objects/stringobject.c?view=markup (lines 1262-1281)